### PR TITLE
Migrating the tracks-callables.js file to the Tracking package.

### PR DESCRIPTION
**Pull request by briancolinger**
Related: 
https://github.com/Automattic/jetpack/issues/15929
https://github.com/Automattic/jetpack/pull/18869

Fixes #

#### Changes proposed in this Pull Request:

* Migrates the tracks-callables.js file to the Tracking package.

#### Does this pull request change what data or activity we track or use?
No Changes

#### Testing instructions:

* Install and activate a 2nd plugin that uses the Jetpack Connection.
* Go to the plugins page on your site
* Verify that w.js and tracks-callables.js are loaded
* Verify the deactivate dialog is presented asking to Deactivate or Disconnect and Deactivate when you deactivate Jetpack.

### CODE REVIEW

**Review by sergeymitr — DISMISSED** [review]
Looks good, tests well 👍 🚢

**Comment by briancolinger** [comment]
Ok, I think this is ready for testing now.
I've grepped the codebase for this file and am not finding any other uses.

File: projects/plugins/jetpack/class.jetpack.php [hunk 1]
```diff
@@ -4149,21 +4149,13 @@ public function deactivate_dialog( $hook ) {
...
-					false
-				);
-
 				wp_enqueue_script(
 					'jetpack-deactivate-dialog-js',
 					Assets::get_file_url_for_environment(
 						'_inc/build/jetpack-deactivate-dialog.min.js',
 						'_inc/jetpack-deactivate-dialog.js'
 					),
-					array( 'jquery', 'jp-tracks-functions' ),
```
Comment (mdbitz) [inline projects/plugins/jetpack/class.jetpack.php:4166]: I'm slightly confused. Shouldn't this dependable still be required? The issue is that it should be enqueued from the package but this jetpack-deactivate-dialog-js still is dependent on it.

Same with jetpack-search-widget-admin below.

Or is this me forgetting how all these wp_register_scripts work.

**Review by mdbitz — CHANGES_REQUESTED** [review]
Need confirmation removal of jp-tracks-functions as a dependency.

File: projects/plugins/jetpack/class.jetpack.php (hunk 1 above)
Comment (briancolinger) [reply projects/plugins/jetpack/class.jetpack.php:4166]: These questions are also relevant to my interests :)
I'm not fully understanding how this all works. Originally, I thought it was as simple as adding this tracks-callable.js file to the Tracking package... but no, there's more to it than that :)

Comment (zinigor) [reply projects/plugins/jetpack/class.jetpack.php:4166]: In my opinion, yes, this dependency should remain here. I guess the original intention behind `wp_register_script` and actual `wp_enqueue_script` was that the first one would only add the script and its path to a list of available files to be loaded, like `jquery`, and then once a script gets enqueued that actually needs that script and lists it as its dependency, it gets added to the output.

We can do something similar here by introducing a new `register_tracks_scripts` function in the Tracks class, and then calling it from packages that need that file. Sort of like this:

```
Tracks::register_scripts();
wp_enqueue_script( ...., array( 'jquery', 'jp-tracks-functions' )
```

That `jp-tracks-functions` name we can also make more verbose to keep it in line with the whole namespacing thing. For example, make it `jetpack-tracks-functions`. Sort of like `plugin-package-file`.

File: projects/plugins/jetpack/class.jetpack.php [hunk 2]
```diff
@@ -4158,13 +4158,8 @@ public function deactivate_dialog( $hook ) {
...
 
-				wp_register_script(
-					'jp-tracks-functions',
-					plugins_url( '_inc/lib/tracks/tracks-callables.js', JETPACK__PLUGIN_FILE ),
-					array( 'jp-tracks' ),
-					JETPACK__VERSION,
-					false
-				);
+				$tracking = new Tracking();
+				$tracking->register_tracks_scripts();
```
Comment (zinigor) [inline projects/plugins/jetpack/class.jetpack.php:4162]: Nicely done! We're making sure that `jp-tracks-functions` is available as a dependency for other scripts. Let's do the same thing in `search.php`?

Comment (briancolinger) [reply projects/plugins/jetpack/class.jetpack.php:4162]: Done.

File: projects/packages/tracking/src/class-tracking.php [hunk 3]
```diff
@@ -87,6 +87,19 @@ public function ajax_tracks() {
 		wp_send_json_success();
 	}
 
+	/**
+	 * Register script necessary for tracking.
+	 */
+	public function register_tracks_scripts() {
```
Comment (mdbitz) [inline projects/packages/tracking/src/class-tracking.php:93]: Lets update this to a static function. There is no reason to need to instantiate a Tracking class.

File: projects/packages/tracking/src/class-tracking.php [hunk 4]
```diff
@@ -99,6 +112,14 @@ public function enqueue_tracks_scripts() {
 			true
 		);
 
+		wp_enqueue_script(
```
Comment (mdbitz) [inline projects/packages/tracking/src/class-tracking.php:115]: jp-tracks-functions should not be enqueued at this point. Doing so will add it to render in the footer regardless of if it is needed.

File: projects/plugins/jetpack/class.jetpack.php [hunk 5]
```diff
@@ -4158,13 +4158,8 @@ public function deactivate_dialog( $hook ) {
...
 				);
 
-				wp_register_script(
-					'jp-tracks-functions',
-					plugins_url( '_inc/lib/tracks/tracks-callables.js', JETPACK__PLUGIN_FILE ),
-					array( 'jp-tracks' ),
-					JETPACK__VERSION,
-					false
-				);
+				$tracking = new Tracking();
```
Comment (mdbitz) [inline projects/plugins/jetpack/class.jetpack.php:4161]: Tracking::register_tracks_scripts();

File: projects/plugins/jetpack/modules/widgets/search.php [hunk 6]
```diff
@@ -136,18 +136,13 @@ public function widget_admin_setup() {
...
 		);
 
-		wp_register_script(
-			'jp-tracks-functions',
-			plugins_url( '_inc/lib/tracks/tracks-callables.js', JETPACK__PLUGIN_FILE ),
-			array(),
-			JETPACK__VERSION,
-			false
-		);
+		$tracking = new Tracking();
```
Comment (mdbitz) [inline projects/plugins/jetpack/modules/widgets/search.php:139]: Tracking::register_tracks_scripts();

File: projects/packages/tracking/src/class-tracking.php [hunk 7]
```diff
@@ -87,6 +87,19 @@ public function ajax_tracks() {
...
 	}
 
+	/**
+	 * Register script necessary for tracking.
+	 */
+	public function register_tracks_scripts() {
+		wp_register_script(
+			'jp-tracks-functions',
+			Assets::get_file_url_for_environment( 'js/tracks-callables.js', 'js/tracks-callables.js', __FILE__ ),
+			array( 'jp-tracks' ),
```
Comment (mdbitz) [inline projects/packages/tracking/src/class-tracking.php:97]: what is jp-tracks?

The track package has jptracks. Is there duplicate or multiple versions. could be a bigger item to untangle.

**Review by mdbitz — CHANGES_REQUESTED** [review]
In addition to the below we also need to expand on the testing instructions. As to when we should have the js file included and ensure dependent functionality works (search widget, etc)

File: projects/packages/tracking/src/class-tracking.php (hunk 7 above)
Comment (mdbitz) [reply projects/packages/tracking/src/class-tracking.php:97]: wp_enqueue_script(
			'jp-tracks',
			'//stats.wp.com/w.js',
			array(),
			gmdate( 'YW' ),
			true
		);

**Comment by mdbitz** [comment]
@zinigor @sergeymitr I've reworked some pieces of this pr to move tracks-callables.js into the Tracks package. Would you both mind reviewing. I am seeing if I can put together Testing Instructions now.

**Review by kbrown9 — APPROVED** [review]
I tested this using the provided testing instructions, and everything worked well for me. The `w.js` and `tracks-callables.js` files were loaded as expected.

Is there a reason that we're leaving the existing `tracks-callables.js` file in the Jetpack plugin?

**Comment by mdbitz** [comment]
No there isn't a reason for the file to still exist in _inc/lib/tracks I've removed it as part of #19126

**Comment by jeherve** [comment]
r228235-wpcom

**Pull request merged**